Fix #77812: Interactive mode does not support PHP 7.3-style heredoc

cmb69 opened the pull request:
As of PHP 7.3.0, the rules regarding the heredoc and nowdoc closing
identifier have been relaxed.  While formerly, the closing identifier
was required to be placed at the beginning of a line and to be
immediately followed by a semicolon and a line break, it may now be
preceeded by whitespace, and may be followed by any non-word character.
We adjust the recognition logic respectively.

carusogabriel commented on ext/readline/tests/bug77812-libedit.phpt:
```diff
@@ -0,0 +1,30 @@
+<?php
+if (!extension_loaded('readline')) die('skip readline extension not available');
+if (READLINE_LIB !== "libedit") die("skip libedit only");
+if (!function_exists('proc_open')) die("proc_open() not available");
```
```suggestion
if (!function_exists('proc_open')) die("skip proc_open() not available");
```

cmb69 replied:
Thanks!

nikic commented on ext/readline/readline_cli.c:
```diff
@@ -349,9 +349,14 @@ static int cli_is_valid_code(char *code, size_t len, zend_string **prompt) /* {{
 			case heredoc:
 				if (code[i - (heredoc_len + 1)] == '\n' && !strncmp(code + i - heredoc_len, heredoc_tag, heredoc_len) && code[i] == '\n') {
```
Can we drop this case?

cmb69 replied:
Good catch, thanks! Indeed, this is now superfluous.

nikic commented on ext/readline/readline_cli.c:
```diff
@@ -349,9 +349,14 @@ static int cli_is_valid_code(char *code, size_t len, zend_string **prompt) /* {{
 					code_type = body;
-				} else if (code[i - (heredoc_len + 2)] == '\n' && !strncmp(code + i - heredoc_len - 1, heredoc_tag, heredoc_len) && code[i-1] == ';' && code[i] == '\n') {
+				} else if (!strncmp(code + i - heredoc_len + 1, heredoc_tag, heredoc_len)) {
+					char c = code[i + 1];
```
Do we need to check string length here?

cmb69 replied:
If we get there, a `strncmp()` starting at `code[i+1-heredoc_len]` has matched for a size of `heredoc_len`, so `code[i+1]` should be always valid.

nikic replied:
The strncmp only inspects memory up to code+i, so I don't think that really tells us anything. There seems to be a bunch of other code using i+1 lookahead though, so I'll assume that these strings are guaranteed null terminated and thus safe to access one past end.

cmb69 replied:
You're right, the `strcmp` doesn't access `code[i+1]`, but there should always be a trailing line break (seems to always be LF even on Windows), so I don't see a problem here.

nikic commented on ext/readline/readline_cli.c:
```diff
@@ -349,9 +349,14 @@ static int cli_is_valid_code(char *code, size_t len, zend_string **prompt) /* {{
+					while (*p == ' ' && *p != '\n') p--;
+					if (*p != '\n') break;
 					code_type = body;
-					valid_end = 1;
```
Why is this removed?

cmb69 replied:
`valid_end` refers to the semicolon here, which is no longer part of `heredoc`.

nikic commented on ext/readline/tests/bug77812-libedit.phpt:
```diff
@@ -0,0 +1,30 @@
+var_dump($proc);
+fwrite($pipes[0], "echo <<<FOO\n    bar\n    FOO;\n");
+fwrite($pipes[0], "print(<<<FOO\nxx\nFOO);\n");
+fwrite($pipes[0], "echo <<<FOO\n    xxx\n    FOO;\nFOO\n;\n");
```
Possibly test FOO with trailing alnum chars?

cmb69 replied:
Makes sense. While trying, I've stumbled upon https://3v4l.org/PGCMI, which is supposed to be valid (should print `FOO42`), from what I can tell.

cmb69 commented:
This PR is blocked by [bug #78441](https://bugs.php.net/bug.php?id=78441).

cmb69 commented:
Rebased onto current PHP-7.3 (to get the fix for [bug # 78441](https://bugs.php.net/bug.php?id=78441)), and added a test with a trailing digit. I've also catered to extended ASCII characters, which are [permissible in (heredoc) labels](https://github.com/php/php-src/blob/310708845ff02b9e252b7a02616290384dd846ba/Zend/zend_language_scanner.l#L117-L118).

nikic commented on ext/readline/readline_cli.c:
```diff
@@ -347,11 +347,14 @@ static int cli_is_valid_code(char *code, size_t len, zend_string **prompt) /* {{
-					code_type = body;
-				} else if (code[i - (heredoc_len + 2)] == '\n' && !strncmp(code + i - heredoc_len - 1, heredoc_tag, heredoc_len) && code[i-1] == ';' && code[i] == '\n') {
+				if (!strncmp(code + i - heredoc_len + 1, heredoc_tag, heredoc_len)) {
+					char c = code[i + 1];
```
I think this might need to be `unsigned char` to make sure that the `>= 0x80` check works correctly on all platforms?

cmb69 replied:
Indeed, you're right. Fixed.

nikic commented on ext/readline/readline_cli.c:
```diff
@@ -347,11 +347,14 @@ static int cli_is_valid_code(char *code, size_t len, zend_string **prompt) /* {{
+					char *p = code + i - heredoc_len;
+
+					if ((c >= 'A' && c <= 'Z') || (c >= 'a' && c <= 'z') || (c >= '0' && c <= '9') || c == '_' || c >= 0x80) break;
+					while (*p == ' ' && *p != '\n') p--;
```
This should also be skipping tabs -- also the last condition is redundant.

cmb69 replied:
Not sure if it's even possible to have tabs, but since other code parts also deal with tabs, it makes sense to support them here as well. Thanks!

cmb69 commented:
Thanks! Applied as 03c7749dc8f40df8f41afffcdef4595ad43afe7f.
